Source/Plugins/Core/com.equella.core/src/com/tle/core/connectors/blackboard/service/impl/BlackboardRESTConnectorServiceImpl.java

```diff
@@ -606,6 +596,56 @@ public boolean supportsReverseSort() {
     }
   }
 
+  private Token getUserSessionAuth() {
+    Optional<Token> auth =
+        userSessionService.getAttribute(BlackboardRESTConnectorConstants.USER_SESSION_AUTH_KEY);
+    if (!auth.isPresent()) {
+      throw new AuthenticationException(
+          "User authentication details are not available in the session");
+    }
+    return auth.get();
+  }
+
+  private void setUserSessionAuth(Optional<Token> t) {
+    userSessionService.setAttribute(BlackboardRESTConnectorConstants.USER_SESSION_AUTH_KEY, t);
+  }
+
+  private void setUserSessionAuth(Token t) {
+    setUserSessionAuth(Optional.fromNullable(t));
+  }
+
+  private void refreshBlackboardToken(Connector connector) {
+    Token expAuth = getUserSessionAuth();
+
+    // Make a reasonable effort to obtain the latest refresh token
+    Optional<Token> cachedAuth = getAuth(connector);
+    if (cachedAuth.isPresent()) {
+      expAuth = cachedAuth.get();
+    }
+
+    final String path =
+        "learn/api/public/v1/oauth2/token?grant_type=refresh_token&refresh_token="
+            + expAuth.getRefreshToken();
+
+    try {
+      // Setting 'firstTime' to false - if the token refresh fails, it's likely not recoverable.
+      setAuth(
+          connector,
+          sendBlackboardData(
+              connector,
+              path,
+              Token.class,
+              "grant_type=refresh_token",
+              Request.Method.POST,
+              false));
+    } catch (Exception exception) {
+      LOGGER.error(exception.getMessage());
+      // Likely an expired auth token that failed to refresh.  Guide the user to try again
+      removeCachedValue(authCache, buildCacheKey(CACHE_ID_AUTH, connector));
+      new AuthenticationException("Unable to refresh auth token.  Please retry action.");
```

Comment:
Is this missing a throw?

```suggestion
      throw new AuthenticationException("Unable to refresh auth token.  Please retry action.");
```

Reply:
doh - yes it is.